### ISSUE TITLE
Tavern 1.7.0 "Hanging"

### ISSUE THREAD

**Issue by KenStorey** [taverntesting/tavern#610]
Tavern intermittently, but frequently, just stops working.

![Tavern1](https://user-images.githubusercontent.com/46034453/97029320-2f021080-1555-11eb-8b00-0d890a0653c0.PNG)
![Tavern 2](https://user-images.githubusercontent.com/46034453/97029453-5f49af00-1555-11eb-87e0-1b8f5fab9f91.PNG)

In the first example it stopped on the first test, in the second it stopped halfway through. Sometimes it even manages to complete the run!

It is running in a Docker container, in this case locally run manually, but the same occurs on our Git pipelines.

The scripts are not very complicated. The first example is

```
paho-mqtt: &mqtt_spec
  client:
    client_id: test_team
  connect:
    host: "{host}"
    port: !int "{port}"
    timeout: 0.5

# test_app_brightness_levels.tavern.yaml

test_name: Brightness, no adjustment

strict:
  - json:off

includes: &includes
  - !include tavern_library/config.yaml
  - !include tavern_library/tdcp_keys.yaml
  - !include tavern_library/hmi_indications.yaml
  - !include tavern_library/mqtt_topics.yaml

paho-mqtt: *mqtt_spec

stages:

  - name: Step 1 - Brightness key press

    mqtt_publish:
      topic: "{dcpsim_keypress_topic}"
      json:
        key_code: !int "{brightnessKey}"

  - name: Step 2 - Brightness key release

    mqtt_publish:
      topic: "{dcpsim_keyrelease_topic}"
      json:
        key_code: !int "{brightnessKey}"

  - name: Step 3 - Brightness indication and level is displayed

    mqtt_publish:
      topic: "{dcpsim_information_request_topic}"
      json:
        request: "current_display"

    mqtt_response:
      topic: "{dcpsim_current_display_text_topic}"
      json:
        line2_text: "{brightness_text}"
        line3_text: "{level4}"
      timeout: 1

  - name: Step 4 - Brightness indication and level are blanked after 3 seconds of inactivity

    delay_before: 3

    mqtt_publish:
      topic: "{dcpsim_information_request_topic}"
      json:
        request: "current_display"

    mqtt_response:
      topic: "{dcpsim_current_display_text_topic}"
      json:
        line2_text: "{blank_line}"
        line3_text: "{blank_line}"
      timeout: 1

```
I attempted to set up logging by reference to [Tavern Documentation](https://tavern.readthedocs.io/en/latest/debugging.html) but I'm afraid I didn't understand it.

**Comment by KenStorey** [comment]
I found this in the MQTT  broker log
```
2020-10-27T15:55:06.136469400Z 1603814106: Received SUBSCRIBE from test_team,
2020-10-27T15:55:06.136510300Z 1603814106: 	hmi/keypressed (QoS 1),
2020-10-27T15:55:06.136800600Z 1603814106: test_team 1 hmi/keypressed,
2020-10-27T15:56:37.003014700Z 1603814197: Client test_team has exceeded timeout, disconnecting.,
2020-10-27T15:56:37.003085100Z 1603814197: Socket error on client test_team, disconnecting.,
```
I've gone back to v1.2.4 and it still occuring.

I never saw it my trusty 1.0.3a build but unfortunately the docker image accidentally got rebuilt and now, instead of running perfectly with a deprecation warning, this happens

```
==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
Direct construction of YamlFile has been deprecated, please use YamlFile.from_parent.
See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.
-------- generated html file: file:///builds/frmcs/hmi/test_report.html --------
=========================== short test summary info ============================
ERROR 
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.08s ===============================

```

**Comment by michaelboulton** [comment]
For the time being if you installed pytest <6 explicitly in your dependencies then I think you should be able to install 1.03a.

I just looked and there hasn't been any changes to the Tavern side of mqtt since 1.0 really, but I did recently change the paho-mqtt library to be `paho-mqtt>=1.3.1,<=1.5.0` when it was previously locked to `paho-mqtt==1.3.1`. I'd suggest explicitly installing that as well.

**Comment by michaelboulton** [comment]
IT might take a while to show up , but I pushed a version 1.11.1 which increases the paho-mqtt version to 1.5.1 which might fix it, I notice they have some bug fixes in their changelog

**Comment by KenStorey** [comment]
Thanks for your assistance.

I rebuilt my old setup with paho-mqtt==1.3.1, pytest==6.0.1 and tavern==1.0.0a3. This worked fine.

I also built with 1.11.1. This seems even worse, hanging on the first test.  There was no disconnection message in the mqtt broker log.

I then locked to paho-mqtt==1.3.1 and installed 1.11.1. This configuration also worked fine so it appears that the later versions of paho are to blame.
